### PR TITLE
Fix analyzer errors in Renci.SshNet and Renci.SshNet.TestTools.OpenSSH

### PR DESCRIPTION
**Pull request by drieseng**
Fix analyzer errors in **Renci.SshNet** and **Renci.SshNet.TestTools.OpenSSH**.
Suppress all errors in unit tests and integration tests.

### CODE REVIEW

File: src/Renci.SshNet/Channels/Channel.cs [hunk 1]
```diff
@@ -158,7 +159,9 @@ public uint RemoteChannelNumber
             {
                 if (!_remoteChannelNumber.HasValue)
                 {
+#pragma warning disable S2372 // Exceptions should not be thrown from property getters
                     throw CreateRemoteChannelInfoNotAvailableException();
+#pragma warning restore S2372 // Exceptions should not be thrown from property getters
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Channels/Channel.cs]: You have disabled S2372 in the .editorconfig, so the suppressions should be unnecessary?

File: src/Renci.SshNet/Common/SshData.cs [hunk 2]
```diff
@@ -146,11 +151,13 @@ protected byte[] ReadBytes()
         /// <exception cref="ArgumentOutOfRangeException"><paramref name="length"/> is greater than the internal buffer size.</exception>
         protected byte[] ReadBytes(int length)
         {
+#pragma warning disable S125 // Sections of code should not be commented out
             /*
              * Note that this also prevents allocating non-relevant lengths, such as if length is greater than _data.Count but less than int.MaxValue.
              * For the nerds, the condition translates to: if (length > data.Count && length < int.MaxValue)
              * Which probably would cause all sorts of exception, most notably OutOfMemoryException.
              */
+#pragma warning restore S125 // Sections of code should not be commented out
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/SshData.cs]: The comment seems out of date

```suggestion
```

File: src/Renci.SshNet/Common/SshData.cs [hunk 3]
```diff
@@ -260,8 +267,12 @@ protected string[] ReadNamesList()
         /// <summary>
         /// Reads next extension-pair data type from internal buffer.
         /// </summary>
-        /// <returns>Extensions pair dictionary.</returns>
+        /// <returns>
+        /// Extensions pair dictionary.
+        /// </returns>
+#pragma warning disable CA1859 // Use concrete types when possible for improved performance
         protected IDictionary<string, string> ReadExtensionPair()
+#pragma warning restore CA1859 // Use concrete types when possible for improved performance
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/SshData.cs]:
```suggestion
        protected Dictionary<string, string> ReadExtensionPair()
```

File: src/Renci.SshNet/Renci.SshNet.csproj [hunk 4]
```diff
@@ -20,4 +20,8 @@
   <PropertyGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' or '$(TargetFramework)' == 'netstandard2.1' or '$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' ">
     <DefineConstants>FEATURE_SOCKET_TAP;FEATURE_SOCKET_APM;FEATURE_SOCKET_EAP;FEATURE_DNS_SYNC;FEATURE_DNS_APM;FEATURE_DNS_TAP</DefineConstants>
   </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Remove="Security\KeyExchangeHash.cs" />
+  </ItemGroup>
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Renci.SshNet.csproj]: Was this on purpose?

File: src/Renci.SshNet/SftpClient.cs [hunk 5]
```diff
@@ -2238,7 +2237,9 @@ private IEnumerable<FileInfo> InternalSynchronizeDirectories(string sourcePath,
                         var remoteFileName = string.Format(CultureInfo.InvariantCulture, @"{0}/{1}", destinationPath, localFile.Name);
                         try
                         {
+#pragma warning disable CA2000 // Dispose objects before losing scope
                             using (var file = File.OpenRead(localFile.FullName))
+#pragma warning restore CA2000 // Dispose objects before losing scope
```
Comment (Rob-Hague) [inline src/Renci.SshNet/SftpClient.cs]: What is the suppression referring to?

**Review by Rob-Hague — COMMENTED** [review]
I still don't see any value in enabling analyzers just to add a bunch of suppressions in the source, but otherwise it is fine for me.

File: src/Renci.SshNet/Common/SshData.cs (hunk 2 above)
Comment (drieseng) [reply src/Renci.SshNet/Common/SshData.cs]: I've remove the comment and added unit tests for this method.

File: src/Renci.SshNet/Common/SshData.cs (hunk 3 above)
Comment (drieseng) [reply src/Renci.SshNet/Common/SshData.cs]: That would be a breaking change. If , but I'm ok with taking the hit on this one.
What do you think?

File: src/Renci.SshNet/Channels/Channel.cs (hunk 1 above)
Comment (drieseng) [reply src/Renci.SshNet/Channels/Channel.cs]: Indeed, remove the suppression from source file.

File: src/Renci.SshNet/Renci.SshNet.csproj (hunk 4 above)
Comment (drieseng) [reply src/Renci.SshNet/Renci.SshNet.csproj]: Nope, removed.

File: src/Renci.SshNet/SftpClient.cs (hunk 5 above)
Comment (drieseng) [reply src/Renci.SshNet/SftpClient.cs]: It's a bug in the roslyn analyzer. Clarified this in latest commit.
I wanted to submit a bug against the roslyn analyzer, but I can't seem to create a small repro for this issue.

File: src/Renci.SshNet/Common/SshData.cs (hunk 3 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/Common/SshData.cs]: It's fine for me, the method is not virtual, so I can't imagine it causing any friction

File: src/Renci.SshNet/Channels/Channel.cs [hunk 6]
```diff
@@ -183,7 +184,9 @@ public uint RemotePacketSize
             {
                 if (!_remotePacketSize.HasValue)
                 {
+#pragma warning disable S2372 // Exceptions should not be thrown from property getters
                     throw CreateRemoteChannelInfoNotAvailableException();
+#pragma warning restore S2372 // Exceptions should not be thrown from property getters
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Channels/Channel.cs]: Sorry for not pointing it out earlier: there are multiple S2372 suppressions that can be removed (probably worth a project-wide search)

File: src/Renci.SshNet/Common/SshData.cs (hunk 3 above)
Comment (drieseng) [reply src/Renci.SshNet/Common/SshData.cs]: Done.

File: src/Renci.SshNet/Channels/Channel.cs (hunk 6 above)
Comment (drieseng) [reply src/Renci.SshNet/Channels/Channel.cs]: Done.

**Review by Rob-Hague — APPROVED** [review]

**Comment by drieseng** [comment]
@WojciechNagorski, can I also get your stamp of approval?

**Review by WojciechNagorski — APPROVED** [review]

**Comment by Rob-Hague** [comment]
Build times and general responsiveness in VS have taken a massive hit. I beg that we reconsider 🥺

**Comment by drieseng** [comment]
@Rob-Hague, I'm perfectly fine with disabling (some of) the analyzers by default. I'll submit a new PR to get the build in an acceptable state.

**Pull request merged**